### PR TITLE
add resolver contract in utils + deploy script + test case base

### PR DESCRIPTION
**Pull request by livingrockrises**
# Summary

resolves V1 and V2 smart account address for given eoa address and range of deployment indexes. for V2 it assumes default auth module is ECDSA module

Related Issue: # (issue number)

## Change Type

- [ ] Bug Fix
- [ ] Refactor
- [ ] New Feature
- [ ] Breaking Change
- [ ] Documentation Update
- [ ] Performance Improvement
- [ ] Other

# Checklist

- [ ] My code follows this project's style guidelines
- [ ] I've reviewed my own code
- [ ] I've added comments for any hard-to-understand areas
- [ ] I've updated the documentation if necessary
- [ ] My changes generate no new warnings
- [ ] I've added tests that prove my fix is effective or my feature works
- [ ] All unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

# Additional Information

# Branch Naming

### CODE REVIEW

File: test/misc/Address-Resolver.specs.ts
Comment (ankurdubey521) [inline test/misc/Address-Resolver.specs.ts:12]: do you want to implement the tests before merging the PR?

Comment (livingrockrises) [reply test/misc/Address-Resolver.specs.ts:12]: PR can wait. will add test and just want to get started with development on sdk using bytecode / test deployments.

File: contracts/smart-account/utils/AddressResolver.sol
Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol:9]: should be fine imo.
the interfaces I am using are not being implemented as they have other methods also.

File: contracts/smart-account/utils/AddressResolver.sol
Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol:22]: done

File: contracts/smart-account/utils/AddressResolver.sol
Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol:10]: if you notice AddressResolver is ! IAddressResolver here. I could write two different interfaces. one being this including struct and other one renamed to ISmartAccountFactory interfaces (cant reuse them as for factory we did not create interfaces)

Comment (ankurdubey521) [reply contracts/smart-account/utils/AddressResolver.sol:10]: > if you notice AddressResolver is ! IAddressResolver here.

Didn't get this

File: contracts/smart-account/utils/AddressResolver.sol
Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol:47]: address(0) will be checked first.  
I thought CREATE2 would return 0 address but I see that's not the case since this is computing without using opcode. 
so read call and anyway for read call this should not matter

File: contracts/smart-account/utils/AddressResolver.sol
Comment (ankurdubey521) [reply contracts/smart-account/utils/AddressResolver.sol:10]: > I could write two different interfaces. one being this including struct and other one renamed to ISmartAccountFactory interfaces (cant reuse them as for factory we did not create interfaces)

Why do we need two different interfaces?

File: contracts/smart-account/utils/AddressResolver.sol
Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol:1]: yeah I can optimise the for loop. overkill for this contract

File: contracts/smart-account/utils/AddressResolver.sol
Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol:10]: AddressResolver does not implement IAddressResolver. which has factory deployment method interface
 
 and factory interface I just figured we have for V2 factory
 before I combined V1 and V2 factory interfaces and named it IAddressResolver which is wrong. 
 
 second interface I mean actual interface where this struct would be moved.
 
 I am pushed updated changes, check them it would make more sense.

Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol:10]: depending on design decisions in the sdk code may change for this contract so my priority is to get mvp out and get it tested. so I would appreciate if rules are not strict - at this stage for such utility contract

Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol]: > AddressResolver does not implement IAddressResolver. which has factory deployment method interface
> 
> and factory interface I just figured we have for V2 factory before I combined V1 and V2 factory interfaces and named it IAddressResolver which is wrong.
> 
> second interface I mean actual interface where this struct would be moved.
> 
> I am pushing updated changes, check them it would make more sense.

https://github.com/bcnmy/scw-contracts/commit/da16499ce8103f43d2a19b759c3095ac4e0d9d22

Comment (ankurdubey521) [reply contracts/smart-account/utils/AddressResolver.sol]: Agreed with keeping things simple, but I think if a change is fairly trivial and helps with consistency and technical debt, we should do the changes.

Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol]: conflict in AddressResolver not implementing interface with same name seemed trivial to me. so I realised and fixed it. In that name some other interface was being used so it makes sense.
moving struct is unnecessary ask imo. anyway I have pushed

**Review by ankurdubey521 — APPROVED** [review]

File: contracts/smart-account/utils/AddressResolver.sol
Comment (filmakarov) [inline contracts/smart-account/utils/AddressResolver.sol:35]: I think this name can be improved to something like `nextArrayElementIndex` which will better reflect the purpose of this var

**Review by filmakarov — APPROVED** [review]
Can be merged after loops index incrementation are unchecked

File: contracts/smart-account/utils/AddressResolver.sol
Comment (livingrockrises) [reply contracts/smart-account/utils/AddressResolver.sol:35]: ok

**Comment by livingrockrises** [comment]
please check @ankurdubey521 @AmanRaj1608 @filmakarov

**Review by Aboudjem — APPROVED** [review]

**Review by AmanRaj1608 — APPROVED** [review]

**Pull request merged**
